src/librustc_mir/transform/const_prop.rs

```diff
@@ -171,6 +171,14 @@ impl<'mir, 'tcx> interpret::Machine<'mir, 'tcx> for ConstPropMachine {
         throw_unsup_format!("calling intrinsics isn't supported in ConstProp");
     }
 
+    fn assert_panic(
+        _ecx: &mut InterpCx<'mir, 'tcx, Self>,
+        _msg: &rustc::mir::interpret::AssertMessage<'tcx>,
+        _unwind: Option<rustc::mir::BasicBlock>,
+    ) -> InterpResult<'tcx> {
+        throw_unsup_format!("panics are not supported in ConstProp");
+    }
```

Comment:
As you can see on line 401, we are catching panics. I'm also very sure that we do see terminators being hit during const prop. Idk how this is not causing test failures. cc @wesleywiser

Reply:
Hm that is curious indeed... line 401 should be dead code with this change, but then tests would fail.

Reply:
Okay it's not dead code. There are still quite a few `throw_panic!` in the codebase.

I think those *should* all be `throw_ub!` as the only actual way to raise a panic from MIR is the `Assert` terminator. But those are also where const-prop gets its diagnostics from...

Reply:
I think what happens is that const-prop is executing dead code (this causes *so many* problems...), otherwise it could not possibly trigger those `throw_panic!`.

It seems like an awful hack to me for const-prop to base its "this expression will panic at runtime" diagnostics not on the actual cause of the panic (the `Assert` terminator), but the UB that gets raised when the panic check is not done properly. I'd prefer to get rid of the "panic" variant of the error enum entirely, do you think that is possible? const_eval could turn the panic into a string already in `assert_panic`, but what would const_prop do?

Reply:
Ideally const_prop would throw those errors in `assert_panic`, turn them into strings there as well, and raise that. They get displayed, after all, so allocation is okay.

I could imagine a `MachineError(String)` variant that the core engine promises to never throw.

Reply:
We don't need to throw anything, we can report the const prop error directly in assert_panic I believe. Since this PR keeps everything working, let's do that cleanup (removing throw_panic) in a separate PR.

About the stuff that should be UB, that may be the source of our duplicate diagnostics problem. Terminators are evaluated in https://github.com/rust-lang/rust/blob/8f1bbd69e13c9e04a4c2b75612bc0c31af972439/src/librustc_mir/transform/const_prop.rs#L839 so the panic thing should indeed become dead code.

Reply:
So the body of `ConstPropMachine::assert_panic`should be `bug!` since we never execute terminators. We may change this in the future to reduce duplication.

r=me with that change

Reply:
>  that may be the source of our duplicate diagnostics problem

Ah, I wasn't just imagining this then.^^

Reply:
> We don't need to throw anything, we can report the const prop error directly in assert_panic I believe.

Oh I see. That makes sense, if we can avoid duplicate execution.

> So the body of ConstPropMachine::assert_panicshould be bug! since we never execute terminators.

That seems to work out.

> Since this PR keeps everything working, let's do that cleanup (removing throw_panic) in a separate PR.

Okay. What is the plan for const_prop here?

Reply:
Let's continue this in https://github.com/rust-lang/rust/issues/66902